hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/supervisor/monitor/ConsumersRuntimeMonitor.java

```diff
@@ -46,44 +50,61 @@ public ConsumersRuntimeMonitor(ConsumersSupervisor consumerSupervisor,
         hermesMetrics.registerGauge("consumers-workload.monitor.oversubscribed", () -> monitorMetrics.oversubscribed);
     }
 
-    public void checkCorrectness() {
-        Set<SubscriptionName> assignedSubscriptions = workloadSupervisor.assignedSubscriptions();
-        Set<SubscriptionName> runningSubscriptions = consumerSupervisor.runningConsumers();
-
-        monitorMetrics.assigned = assignedSubscriptions.size();
-        monitorMetrics.running = runningSubscriptions.size();
-
-        Set<SubscriptionName> missingSubscriptions = missing(assignedSubscriptions, runningSubscriptions);
-        monitorMetrics.missing = missingSubscriptions.size();
-        for (SubscriptionName subscriptionName : missingSubscriptions) {
+    public void log(Set<SubscriptionName> assigned,
+                    Set<SubscriptionName> running,
+                    Set<SubscriptionName> missing,
+                    Set<SubscriptionName> oversubscribed) {
+        for (SubscriptionName subscriptionName : missing) {
             logger.warn("Missing consumer process for subscription: {}", subscriptionName);
         }
 
-
-        Set<SubscriptionName> oversubscribedSubscriptions = oversubscribed(assignedSubscriptions, runningSubscriptions);
-        monitorMetrics.oversubscribed = oversubscribedSubscriptions.size();
-        for (SubscriptionName subscriptionName : oversubscribedSubscriptions) {
+        for (SubscriptionName subscriptionName : oversubscribed) {
             logger.warn("Unwanted consumer process for subscription: {}", subscriptionName);
         }
 
         logger.info(
                 "Subscriptions assigned: {}, existing subscriptions: {}, missing: {}, oversubscribed: {}",
-                assignedSubscriptions.size(),
-                runningSubscriptions.size(),
-                missingSubscriptions.size(),
-                oversubscribedSubscriptions.size()
+                assigned.size(),
+                running.size(),
+                missing.size(),
+                oversubscribed.size()
         );
     }
 
     @Override
     public void run() {
         try {
-            checkCorrectness();
+            Set<SubscriptionName> assigned = workloadSupervisor.assignedSubscriptions();
+            Set<SubscriptionName> running = consumerSupervisor.runningConsumers();
+            Set<SubscriptionName> missing = missing(assigned, running);
+            Set<SubscriptionName> oversubscribed = oversubscribed(assigned, running);
+
+            log(assigned, running, missing, oversubscribed);
+            updateMetrics(assigned, running, missing, oversubscribed);
+
+            ensureCorrectness(missing, oversubscribed);
```

Comment:
maybe we should print some warnings that monitor is fixing runtime tree, otherwise we might not spot long term issues